File: TOKI/tlx/Bab 5/5A - Dichotomic Search/Hadiah.cpp
#include <algorithm>
#include <cstdio>
using namespace std;
int main ()
{
	long long a,b,c,hsl,temp,x,i,j,n;
	long long biner[100],bin[100],list[100];
	scanf("%lld%lld%lld%lld",&a,&b,&c,&n);
	if (a==0)
	{
		printf("1\n");
		return 0;
	}
	if (n==1)
	{
		printf("1\n");
		return 0;
	}
	if (b==0)
	{
		if (n==1)
		{
			printf("1\n");
			return 0;
		}
		if (n>1)
		{
			printf("2\n");
			return 0;
		}
	}
	if (a==1)
	{
		printf("2\n");
		return 0;
	}
	if (c==0)
	{
		hsl=((a%n)+1);
		printf("%lld\n",hsl);
		return 0;
	}
	if (b==1)
	{
		hsl=((a%n)+1);
		printf("%lld\n",hsl);
		return 0;
	}
	x=1;
	while (b!=1)
	{
		biner[x]=b%2;
		b/=2;
		x+=1;
	}
	biner[x]=1;
	for (i=1;i<=x;i++)
	{
		bin[i]=biner[x-i+1];
	}
	temp= (a%n);
	for (i=1;i<=c;i++)
	{
		list[1]=temp;
		for (j=2;j<=x;j++)
		{
			list[j]=list[j-1]*list[j-1];
			list[j]=(list[j]%n);
			if (bin[j]==1)
			{
				list[j]*=temp;
				list[j]%=n;
			}
		}
	temp=list[x];
	}
	hsl=temp+1;
	printf("%lld\n",hsl);
	return 0;
}

File: TOKI/tlx/Bab 5/5A - Dichotomic Search/Hadiah.pas
var
	biner,bin:array[0..30] of byte;
	list:array[0..30] of int64;
	a,b,c,n,x,num,temp,hsl:int64;
	i,dummy:longint;
begin
	readln(a,b,c,n);
	if n=1 then
		begin
		writeln('1');
		exit;
		end;
	if a=1 then
		begin
		writeln('1');
		exit;
		end;		
	if a=0 then
		begin
		writeln('1');
		exit;
		end;
	if b=0 then
		begin
		writeln('2');
		exit;
		end;
	if c=0 then
		begin
		hsl:=((a mod n)+1);
		writeln(hsl);
		exit;
		end;
	x:=1;
	while b>0 do
		begin
		biner[x]:=(b mod 2);
		b:=b div 2;
		x:=x+1;
		end;
	for i:=1 to x-1 do
		begin
		bin[i]:=biner[x-i];
		end;
	temp:=(a mod n);
	x:=x-1;
	for dummy:=1 to c do
		begin
		list[1]:=(temp mod n)*(temp mod n);
		list[1]:=list[1] mod n;
		if bin[i]=1 then
			begin
			list[1]:=list[1]*temp;
			list[1]:=list[1] mod n;
			end;
		for i:=2 to x do
			begin
			num:=list[i-1]*list[i-1];
			num:=num mod n;
			if bin[i]=1 then
				begin
				num:=num*temp;
				end;
			num:=num mod n;
			list[i]:=num;
			end;
		temp:=list[x] mod n;
		end;
	temp:=temp+1;
	writeln(temp);
end.